### DISCUSSION TITLE
how web works and printing

### DISCUSSION THREAD

**Discussion by Fxplorer** [AvaloniaUI/Avalonia#14898] [General]
I am still figuring out how Avalonia works and flows through it processes. I understanding that printing from desktop is more complicated and might need to go through the PDF path. [AvaloniaUI.PrintToPDF](https://github.com/Oaz/AvaloniaUI.PrintToPDF)

I did an xplot template setup and run the browser version fully expecting when I print in the browser window that the page would print (i.e. "Welcome to Avalonia!"). To my surprise, this is not the case. 

**Explaining to me like I am 5, how is Avalonia pushing the screen through in the browser and yet the print will not pick it up?** 

Buttons and text boxes, etc will interact in there. I know you have to use `SelectableTextBlock` instead of `TextBlock` to allow highlighting and copy the text to work. I thought well maybe a 'SelectableTextBlock' will print, but that still results in not printing. 

I tried the same in Blazor Webassembly app and it does print there. I tried this because Blazor is running through WASM in the browser also and I thought that maybe it was because of that.

**Comment by maxkatz6** [comment]
> Explaining to me like I am 5, how is Avalonia pushing the screen through in the browser and yet the print will not pick it up?

Well. That's simple - some browser might not capture html-canvas content. 
Edge/chromium, though, doesn't seem to have this problem:
![image](https://github.com/AvaloniaUI/Avalonia/assets/3163374/f31f60c0-600e-464c-9e83-5d9b785dff56)

> I tried the same in Blazor Webassembly app and it does print there. 

Blazor uses HTML tree, not canvas rendering like Avalonia.

**Reply by maxkatz6** [reply]
In general, for both desktop and browser I would recommend saving controls to PDF and then saving it with file picker or something like this (especially on browser, where you simply can't use c# File API).

**Reply by maxkatz6** [reply]
Checked [AvaloniaUI.PrintToPDF](https://github.com/Oaz/AvaloniaUI.PrintToPDF), and it seem to overuse reflection to get internal APIs.
Just sent a PR there to use public APIs: https://github.com/Oaz/AvaloniaUI.PrintToPDF/pull/4

**Reply by Fxplorer** [reply]
Thank you for this PR and the information provided in it!

**Comment by Fxplorer** [comment]
So the Avalonia "engine" uses Skia to draw what one sees and pushes that to the screen. Skia has the ability to put that into a PDF as well. "recommend saving controls to PD" is referring to that process? 
PrintToPDF references Skia and a "Visual Tree" from Avalonia, so my understanding is that is what it is doing. 
I am referencing this [image ](https://avaloniaui.net/img/how-it-works/Architecture.png) and this discussion #[13077](https://github.com/AvaloniaUI/Avalonia/discussions/13077) as well.

**Reply by maxkatz6** [reply]
Yes. PrintToPDF is fine, but you essentially only need these couple of lines to render any control:
```csharp
public async Task Print(string fileName, Visual control)
{
    using var doc = SKDocument.CreatePdf(fileName);
    var bounds = control.Bounds;
    var page = doc.BeginPage((float)bounds.Width, (float)bounds.Height);
    await DrawingContextHelper.RenderAsync(page, visual);
    doc.EndPage();
}
```

`DrawingContextHelper` is a skia based helper that can output rendering into the skia SKCanvas (pdf page in this context).

**Reply by Fxplorer** [reply]
This is awesome information! Thank you.

**Comment by Fxplorer** [comment]
I think your solitaire game is doing something different. If you create an xplot template and run the browser project, none of the following browsers showed the canvas:
Edge, Chrome, Brave, Firefox 
on Windows 10 machine

I searched around and found Print.js and this [Fiddle ](https://jsfiddle.net/xqsv5hu7/1/) from this [stackoverflow](https://stackoverflow.com/questions/923885/capture-html-canvas-as-gif-jpg-png-pdf) and looked at the [w3shools canvas pages](https://www.w3schools.com/html/html5_canvas.asp) and tried to put the script in the index.html page. It seems the canvas context is not in reach of the Print.js script and it seems the id of the canvas changes each run so I was not able to grab it. So no luck printing(PDF) with this hack.  
In the post [DESKTOP APPS WITH AVALONIA UI AND F#](https://lostindetails.com/articles/Desktop-Apps-with-AvaloniaUI-and-FSharp) of the [LOST IN DETAILS](https://lostindetails.com/) blog near the bottom is the section of "**MISSING PREVIEW / DESIGNER**". Here he is getting images, but it seems to also be from Skia. The articale must be older because of the versions of dotnet and Avalonia, but the ideas seems to all point to the Skia method to get what we want.

**Reply by maxkatz6** [reply]
>  If you create an xplot template and run the browser project, none of the following browsers showed the canvas

Not sure what you mean. This canvas is created dynamically during the startup.
Note, it's not just solitaire, any Avalonia app works this way. It is rendered into HTML Canvas element. 

<img width="842" alt="Screenshot 2024-03-12 at 11 22 28 PM" src="https://github.com/AvaloniaUI/Avalonia/assets/3163374/5a5544ab-2a91-486a-8bec-7c00123c00f5">

**Reply by Fxplorer** [reply]
What I mean is that if you create a new xplot from template, run the browser project and then hit CTRL+P to bring up the print dialog, the canvas is not showing up. The print page should be showing the canvas with the "Welcome to Avalonia" in the center of the page and it is not. Doing so on other pages (websites) with a canvas on the page will print the canvas. There is something happening where the canvas from Avalonia is not getting picked up.

**Reply by Fxplorer** [reply]
OK, you can dis-regard this part of the question. I tried to read up a bit on Skia and the page at Skia [CanvasKit - Quickstart](https://skia.org/docs/user/api/skpaint_overview/) also does not show the canvas in the browser print. So I am just accepting this fact. It just surprised me. I did also learn that even Chrome is actually using Skia to output the PDF and then prints that PDF. I learned some things along the journey, but the simple conclusion is that to print (or generate an image too), one needs to run it through Skia and then do something with the output. Thank you for humoring me!
